core/finality-grandpa/src/until_imported.rs

```diff
@@ -0,0 +1,565 @@
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper stream for waiting until one or more blocks are imported before
+//! passing through inner items. This is done in a generic way to support
+//! many different kinds of items.
+//!
+//! This is used for votes and commit messages currently.
+
+use super::{BlockStatus, Error, SignedMessage, CompactCommit};
+
+use client::ImportNotifications;
+use futures::prelude::*;
+use futures::stream::Fuse;
+use parking_lot::Mutex;
+use runtime_primitives::traits::{Block as BlockT, Header as HeaderT, NumberFor};
+use substrate_primitives::AuthorityId;
+use tokio::timer::Interval;
+
+use std::collections::{HashMap, VecDeque};
+use std::sync::{atomic::{AtomicUsize, Ordering}, Arc};
+use std::time::{Duration, Instant};
+
+// something which will block until imported.
+pub(crate) trait BlockUntilImported<Block: BlockT>: Sized {
+	// the type that is blocked on.
+	type Blocked;
+
+	/// new incoming item. For all internal items,
+	/// check if they require to be waited for.
+	/// if so, call the `Wait` closure.
+	/// if they are ready, call the `Ready` closure.
+	fn schedule_wait<S, Wait, Ready>(
+		input: Self::Blocked,
+		status_check: &S,
+		wait: Wait,
+		ready: Ready,
+	) -> Result<(), Error> where
+		S: BlockStatus<Block>,
+		Wait: FnMut(Block::Hash, Self),
+		Ready: FnMut(Self::Blocked);
+
+	/// called when the wait has completed. The canonical number is passed through
+	/// for further checks.
+	fn wait_completed(self, canon_number: NumberFor<Block>) -> Option<Self::Blocked>;
+}
+
+/// Buffering imported messages until blocks with given hashes are imported.
+pub(crate) struct UntilImported<Block: BlockT, Status, I, M: BlockUntilImported<Block>> {
+	import_notifications: Fuse<ImportNotifications<Block>>,
+	status_check: Status,
+	inner: Fuse<I>,
+	ready: VecDeque<M::Blocked>,
+	check_pending: Interval,
+	pending: HashMap<Block::Hash, Vec<M>>,
+}
+
+impl<Block: BlockT, Status, I: Stream, M> UntilImported<Block, Status, I, M>
+	where Status: BlockStatus<Block>, M: BlockUntilImported<Block>
+{
+	/// Create a new `UntilImported` wrapper.
+	pub(crate) fn new(
+		import_notifications: ImportNotifications<Block>,
+		status_check: Status,
+		stream: I,
+	) -> Self {
+		// how often to check if pending messages that are waiting for blocks to be
+		// imported can be checked.
+		//
+		// the import notifications interval takes care of most of this; this is
+		// used in the event of missed import notifications
+		const CHECK_PENDING_INTERVAL: Duration = Duration::from_secs(5);
```

Comment:
Why not just a `let` binding and a lower case variable name?

Reply:
it seems more idiomatic to use constant bindings for constants